plugins_src/autouv/wpc_snap_win.erl

```diff
@@ -0,0 +1,759 @@
+    St1#st{mat=NewMats#st.mat}.
+
+dup_mat(MatName,{Used,St0},{Image,Name}) ->
+    Mat0 = gb_trees:get(MatName, St0#st.mat),
+    Maps0 = proplists:get_value(maps, Mat0),
+    case proplists:get_value(diffuse, Maps0) of
+	Image ->
+	    %% It already has the texture; no need to create new material
+	    {MatName,{[{MatName,MatName}|Used],St0}};
+	Else ->
+	    NewMatName = list_to_atom(atom_to_list(MatName) ++ "_" ++ Name),
+	    Maps = case Else of
+		       undefined -> [{diffuse,Image}|Maps0];
+		       _ -> lists:keyreplace(diffuse,1,Maps0,{diffuse,Image})
+		   end,
+
+	    %% Make sure that vertex colors will not override the texture.
+	    %% If the Vertex Color attribute in the material is 'Set',
+	    %% change it to 'Multiply'.
+	    OpenGL0 = proplists:get_value(opengl, Mat0),
+	    VtxColors = case proplists:get_value(vertex_colors, OpenGL0) of
+			    set -> multiply;
+			    Other -> Other
+			end,
+	    OpenGL1 = proplists:delete(vertex_colors, OpenGL0),
+	    OpenGL = [{vertex_colors,VtxColors}|OpenGL1],
+
+	    Mat1 = proplists:delete(opengl, Mat0),
+	    Mat2 = proplists:delete(maps, Mat1),
+	    Mat = {NewMatName,[{opengl,OpenGL},{maps,Maps}|Mat2]},
+	    case wings_material:add_materials([Mat], St0) of
+		{St,[]} ->
+		    {NewMatName, {[{MatName,NewMatName}|Used],St}};
+		{St,[{NewMatName,ChangedMatName}]} ->
+		    {ChangedMatName, {[{MatName,ChangedMatName}|Used],St}}
+	    end
+    end.
+
+
+%% win_data/1 function allows many plugin windows to be saved.
+%% it returns: {Name, {Horiz alignment, Custom_data}}
+%% horiz alignment should be either "left" or "right"
+%% custom data is used to store windows properties and custom data - it should be parsed in window/5
+win_data(?WIN_NAME=Name) ->
+    {Name, {right,[]}}.
+
+win_name() ->
+    ?WIN_NAME.
+
+title() ->
+    ?__(1,"Snap Image").
+
+window(St) ->
+    case wings_wm:is_window(?WIN_NAME) of
+	true ->
+	    wings_wm:raise(?WIN_NAME),
+	    keep;
+	false ->
+	    wpc_snap_win:window(?WIN_NAME, {5,150}, {-1,-1}, [], St),
+	    keep
+    end.
+
+window(Name, Pos, Size, Ps0, St) ->
+    State = get_state(Name),
+    {Frame,Ps} = wings_frame:make_win(title(), [{size, Size}, {pos, Pos}|Ps0]),
+    Window = wx_object:start_link(?MODULE, [Frame, Name, State], []),
+    wings_wm:toplevel(Name, Window, Ps, {push,change_state(Window, St)}),
+    keep.
+
+%%%%%%%% Window internals %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Inside wings (process)
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+change_state(Window, St) ->
+    fun(Ev) -> forward_event(Ev, Window, St) end.
+
+forward_event(redraw, _Window, _St) -> keep;
+forward_event(init_opengl, _Window, _St) -> keep;
+forward_event({apply, ReturnSt, Fun}, Window, St0) ->
+    %% Apply ops from window in wings process
+    case ReturnSt of
+	true ->
+	    St = Fun(St0),
+	    {replace, change_state(Window, St)};
+	false ->
+	    Fun(St0)
+    end;
+forward_event(Ev, Window, _) ->
```

Comment:
Do you need to forward all events here?
It's better to only forward the events you need, so that we don't send #st{} or #we{} because memory
will explode if we dot that.